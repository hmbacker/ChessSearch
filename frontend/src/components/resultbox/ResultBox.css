.result-box {
  border-collapse: collapse;
  max-width: 1200px;
  margin: auto;
}

.result-box a:link {
  color: skyblue;
}

.result-box a:visited {
  color: skyblue;
}

.result-box a:hover {
  color: rgb(191, 237, 255);
}

.result-row td, .result-header-cell {
  border: 2px groove white;
  text-align: left;
  padding: 2px 4px;
}

.result-box-header {
  font-weight: bold;
}

.result-header-cell:hover {
  cursor: pointer;
}

@keyframes rotating {
  from {
    transform: rotate(360deg);
  }
  to {
    transform: rotate(0deg);
  }
}

.loading-symbol-cell {
  text-align: center;
  padding: 10px;
}

.loading-symbol-cell img {
  display: inline-block;
  max-width: 50px;
  animation: rotating 0.5s linear infinite;
}
