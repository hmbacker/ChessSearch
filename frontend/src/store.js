// This is the store! We likely will not have to change this file throughout the whole project :)

import { createStore } from "redux";
import rootReducer from "./reducers/index";

const store = createStore(rootReducer);

export default store;
