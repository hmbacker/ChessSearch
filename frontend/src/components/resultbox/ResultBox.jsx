import React, { Component } from "react";
import { fetchSearchResults, fetchMoreSearchResults } from "../../actions";
import { connect } from "react-redux";
import "./ResultBox.css";
import ResultRow from "./ResultRow";
import ResultHeaderCell from "./ResultHeaderCell";
import loadingSvg from "../../loading.svg";

const fetchMoreTriggerDistance = 150;

class ResultBox extends Component {
  constructor(props) {
    super(props);

    this.handleHeaderClick = this.handleHeaderClick.bind(this);
    this.handleScroll = this.handleScroll.bind(this);
  }

  handleHeaderClick(headerField) {
    // Order by the field corresponding to the header clicked.
    if (Object.keys(this.props.mostRecentParamObject).length === 0) {
      // There is no "most recent param object", so we have no search to sort.
      return;
    }

    const paramObject = Object.assign({}, this.props.mostRecentParamObject, {
      offset: 0,
      limit: 25
    });

    if (this.props.mostRecentParamObject.orderby === headerField) {
      // Already ordered by headerField, so switch between ascending/descending
      paramObject.descending = !paramObject.descending;
    } else {
      // Not already ordered by headerField, make it ordered by headerField and ascending
      paramObject.orderby = headerField;
      paramObject.descending = false;
    }

    // Fetch new results with this ordering
    this.props.fetchSearchResults(paramObject);
  }

  handleScroll() {
    // Fetch more search results if we have scrolled to the bottom of the screen
    const root = document.documentElement;
    if (
      window.innerHeight + window.scrollY <
      root.scrollHeight - fetchMoreTriggerDistance
    ) {
      return;
    }
    this.props.fetchMoreSearchResults(25);
  }

  componentDidMount() {
    window.addEventListener("scroll", this.handleScroll);
  }

  render() {
    const titlesAndFieldNames = [
      ["White Player", "white_id"],
      ["Rating", "white_rating"],
      ["Black Player", "black_id"],
      ["Rating", "black_rating"],
      ["Winner", "winner"],
      ["Turns", "turns"],
      ["Opening", "opening_eco"],
      ["Increment", "increment_code"],
      ["Time", "last_move_at"]
    ];
    return (
      <table className="result-box">
        <thead className="result-box-header">
          <tr>
            {titlesAndFieldNames.map(tuple => (
              <ResultHeaderCell
                key={tuple[1]}
                title={tuple[0]}
                fieldName={tuple[1]}
                paramObject={this.props.mostRecentParamObject}
                handleHeaderClick={this.handleHeaderClick}
              />
            ))}
          </tr>
        </thead>
        <tbody>
          {this.props.results.map(game => {
            return <ResultRow game={game} key={game.id} />;
          })}
          <tr>
            <td className="loading-symbol-cell" colSpan="10">
              {// If we are fetching data, show a spinny loading symbol
              this.props.isFetching ? (
                <img src={loadingSvg} alt="Loading"></img>
              ) : null}
            </td>
          </tr>
        </tbody>
      </table>
    );
  }
}

function mapStateToProps(state) {
  return {
    isFetching: state.searchResults.isFetching,
    mostRecentParamObject: Object.assign(
      {},
      state.searchResults.mostRecentParamObject
    ),
    results: state.searchResults.results
  };
}

function mapDispatchToProps(dispatch) {
  return {
    fetchSearchResults: paramObject =>
      dispatch(fetchSearchResults(paramObject)),
    fetchMoreSearchResults: limit => dispatch(fetchMoreSearchResults(limit))
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ResultBox);
