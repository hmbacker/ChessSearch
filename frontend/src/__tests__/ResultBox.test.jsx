import React from "react";
import ResultBox from "../components/resultbox/ResultBox";
import { shallow } from "enzyme";
import store from "../store";

test("renders correctly", () => {
  const wrapper = shallow(<ResultBox store={store} />);
  expect(wrapper.debug()).toMatchSnapshot();
});
