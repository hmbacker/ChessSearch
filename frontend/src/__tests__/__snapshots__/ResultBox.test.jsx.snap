// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
"<ContextProvider value={{...}}>
  <ResultBox store={{...}} isFetching={false} mostRecentParamObject={{...}} results={{...}} fetchSearchResults={[Function: fetchSearchResults]} fetchMoreSearchResults={[Function: fetchMoreSearchResults]} />
</ContextProvider>"
`;
